### ISSUE TITLE
The heat exchanger model in basic modelica library is not working

### ISSUE THREAD

**Issue by Tamim-H0ssain** [OpenModelica/OpenModelica#9914]
### Description

Hi,
When I try to use the heat exchanger model in Modelica.Fluid.Examples.HeatExchanger.BaseClasses.BasicHX, I get the following error 
```
Simulation process failed. Exited with code 255.
Failed to solve the initialization problem with global homotopy with equidistant step size.
Unable to solve initialization problem.
simulation terminated by an assertion at initialization
```
I have also tried the example given in the Modelica.Fluid.Examples.HeatExchanger.HeatExchangerSimulation directory but It does not work either. I also tried to use it with ThermoPower library components, for some input values it gives
```
Matrix singular!
under-determined linear system not solvable!
Warning: maximal number of iteration reached but no root found
```

### Version and OS

 - OpenModelica Version: 1.20.0
 - OS: Windows 10, 64 bit
 - Modelica library version: 4.0.0

**Comment by casella** [comment]
> @casella Not a single example given in the standard modelica library or thermo power library with a heat exchanger does run.

This is not true. If you check ThermoPower.Examples.CISE.Simulators, ThermoPower.Examples.HRB.Simulators, or ThermoPower.Examples.RankineCycle.Simulators, you'll see several examples of simple power plant models with heat exchangers that indeed do run. At least this is what I see from the latest [OpenModelica ThermoPower report](https://libraries.openmodelica.org/branches/master/ThermoPower/ThermoPower.html).

> There are numerous plant examples with HE and do not run. Not a single one. Maybe I am doing sth wrong. For example Modelica.fluids.Examples.HeatExchanger.HeatExchangerSimulation. It is very frustrating.

I understand your frustration. In fact I had my fair share of it when I began doing these things. That's why I've been working on research topics to alleviate it, such as homotopy or advanced Modelica model debugging. There is still a lot of work to be done in that area.

The problem is that thermo-hydraulic dynamic system simulation is difficult. Modelica takes away the pain of programming the models from the grounds up, which is great; I did that during my PhD and it was not fun. But it doesn't take away the complexity of the system behaviour. In fact, it makes it very easy to write down very involved models that are almost impossible to get going numerically. I was very lucky to have an exceptional guide when I started working on this, my PhD supervisor prof. Claudio Maffezzoni. I learned most of what I know in this area from him. Unfortunately he got very busy in a number of different areas, so he never managed to write a textbook on the topic, and even more unfortunately he passed away prematurely in 2005. I'd like to write a book on it myself, based on what I learned in the last 20 years playing with these things. That would need taking a year off, not sure when I'll be able to do that 😅 

First of all, thermo-hydraulic models can be very different depending on the modelling assumptions (compressible/incompressible fluid, liquid, ideal/non-ideal gas, etc).  You have coupled nonlinear mass, momentum, and energy equations, as well as nonlinear constitutive equations such as heat transfer or friction correlations. Everything is horrendously nonlinear. Plus, if you are not careful, you get flow reversal, which is a nightmare from a modelling point of view. Not to mention two-phase flow and heat transfer, another source of strongly nonlinear behaviour. Last, but not least, if you want to initialize systems in steady state, you need to solve initial equations that are horrendously nonlinear, and this may be problematic. I guess that is the problem with Modelica.Fluid.Examples.HeatExchanger.HeatExchangerSimulation. You can try to switch from steady-state initialization to fixed initial state initialization, but then you may end up with extremely weird transients, possibly causing flow reversal, that may be equally difficult to handle.

There are strategies and tricks to cope with this, such as using homotopy, see e.g. [this paper](http://dx.doi.org/10.3384/ecp1106386), and also first building and testing individual heat exchangers and then gradually connecting them one at a time until you have the full plant model. But the whole thing is an art. Forget about plug-and-play thermo-hydraulic simulation.

In fact, my friend Dirk Zimmer has been struggling with this fact for many years, and he came up with a different approach, which is quite smart. It trades off some modelling accuracy for increased robustness. Apparently his approach is more likely to work in a plug-and-play fashion than traditional models such as you find in ThermoPower or Modelical.Fluid. 

You can check his ThermoFluidStream library and the paper he wrote about it. ThermoFluidSystems is supported by OpenModelica.

> Can you please lead to an example of heat exchanger that works?

See above. In fact, ThermoPower.PowerPlants was meant to contain many such examples. Unfortunately, we changed some interfaces some years ago, and I never had the resources and time to fully update it to the new interfaces, so those models are still broken. Sorry about that.

**Comment by Akhlak69** [comment]
@casella  Thanks for making time and giving me such a nice and elaborate reply. Feels so good to get help in OpenModelica, the community is so small (as it seems to me). Thanks for sharing the paper and the new library, I will definitely go through them. 

Please start writing the book, maybe with little steps. Will be very helpful for new users like me.

I am also doing my PhD, I already made an Integrated Energy System model by slowly connecting component by component. I enjoyed it, I am now more used to it, I do a lot of stuffs that others do in python. But now all the problem comes down to a heat exchanger.

I tried all those CISE, Rankine and HRB simulators several times before and again tried running them after getting your message and always got the following error. I did not change anything or gave any inputs, just ran the examples.

Compilation process failed. Exited with code ox00000002.

If you have time can you run any one them you suggested and see if you get the error?

**Comment by Akhlak69** [comment]
Thanks a lot. 

I am using OpenModelica 1.18.1 in Windows. 

I tried running the HRB closed loop simulators by adding thermo power system library and modelica fluid system library, still shows compilation error. 

Not sure how to do it but I will try "use the master version of the library directly from a local git repo following the one on GitHub" and let you know.

**Comment by casella** [comment]
> Thanks a lot.
> 
> I am using OpenModelica 1.18.1 in Windows.

OMG, why do you use such an old version? You should use 1.21.0. There's no reason whatsoever to continue using 1.18.1.
